process/block/preprocess/smartContractResults.go

```diff
@@ -0,0 +1,424 @@
+		onRequestSmartContractResult: onRequestSmartContractResult,
+		scrProcessor:                 scrProcessor,
+		accounts:                     accounts,
+	}
+
+	scr.chRcvAllScrs = make(chan bool)
+	scr.scrPool.RegisterHandler(scr.receivedSmartContractResult)
+	scr.scrForBlock.txHashAndInfo = make(map[string]*txInfo)
+
+	return scr, nil
+}
+
+// waitForScrHashes waits for a call whether all the requested smartContractResults appeared
+func (scr *smartContractResults) waitForScrHashes(waitTime time.Duration) error {
+	select {
+	case <-scr.chRcvAllScrs:
+		return nil
+	case <-time.After(waitTime):
+		return process.ErrTimeIsOut
+	}
+}
+
+// IsDataPrepared returns non error if all the requested smartContractResults arrived and were saved into the pool
+func (scr *smartContractResults) IsDataPrepared(requestedScrs int, haveTime func() time.Duration) error {
+	if requestedScrs > 0 {
+		log.Info(fmt.Sprintf("requested %d missing scr\n", requestedScrs))
+		err := scr.waitForScrHashes(haveTime())
+		scr.scrForBlock.mutTxsForBlock.RLock()
+		missingScrs := scr.scrForBlock.missingTxs
+		scr.scrForBlock.mutTxsForBlock.RUnlock()
+		log.Info(fmt.Sprintf("received %d missing scr\n", requestedScrs-missingScrs))
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// RemoveTxBlockFromPools removes smartContractResults and miniblocks from associated pools
+func (scr *smartContractResults) RemoveTxBlockFromPools(body block.Body, miniBlockPool storage.Cacher) error {
+	if body == nil {
+		return process.ErrNilTxBlockBody
+	}
+
+	err := scr.removeDataFromPools(body, miniBlockPool, scr.scrPool, block.SmartContractResultBlock)
+
+	return err
+}
+
+// RestoreTxBlockIntoPools restores the smartContractResults and miniblocks to associated pools
+func (scr *smartContractResults) RestoreTxBlockIntoPools(
+	body block.Body,
+	miniBlockHashes map[int][]byte,
+	miniBlockPool storage.Cacher,
+) (int, error) {
+	if miniBlockPool == nil {
+		return 0, process.ErrNilMiniBlockPool
+	}
+
+	scrRestored := 0
+	for i := 0; i < len(body); i++ {
+		miniBlock := body[i]
+		if miniBlock.Type != block.SmartContractResultBlock {
+			continue
+		}
+
+		strCache := process.ShardCacherIdentifier(miniBlock.SenderShardID, miniBlock.ReceiverShardID)
+		scrBuff, err := scr.storage.GetAll(dataRetriever.SmartContractResultUnit, miniBlock.TxHashes)
+		if err != nil {
+			return scrRestored, err
+		}
+
+		for txHash, txBuff := range scrBuff {
+			tx := smartContractResult.SmartContractResult{}
+			err = scr.marshalizer.Unmarshal(&tx, txBuff)
+			if err != nil {
+				return scrRestored, err
+			}
+
+			scr.scrPool.AddData([]byte(txHash), &tx, strCache)
+		}
+
+		err = scr.restoreMiniBlock(miniBlock, miniBlockPool, miniBlockHashes[i])
+		if err != nil {
+			return scrRestored, err
+		}
+
+		scrRestored += len(miniBlock.TxHashes)
+	}
+
+	return scrRestored, nil
+}
+
+// ProcessBlockSmartContractResults processes all the smartContractResult from the block.Body, updates the state
```

Comment:
Comment typo, it should be: "ProcessBlockTransactions"

Reply:
done